File: 718.cpp
/*Given two integer arrays A and B, return the maximum length of an subarray that appears in both arrays.

Example 1:
Input:
A: [1,2,3,2,1]
B: [3,2,1,4,7]
Output: 3
Explanation: 
The repeated subarray with maximum length is [3, 2, 1].
Note:
1 <= len(A), len(B) <= 1000
0 <= A[i], B[i] < 100
*/

class Solution {
public:
    
    // dp(i,j) = A[i-1] == B[j-1] ? dp[i-1][j-1] + 1 : 0;
    int findLength(vector<int>& A, vector<int>& B) {
        int m = A.size(), n = B.size();
        // vector<vector<int>> dp(m+1, vector<int>(n+1, 0));
        vector<int> pre(n+1, 0), cur(n+1, 0);
        int res = 0;
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (A[i-1] == B[j-1]) cur[j] = pre[j-1] + 1;
                else cur[j] = 0;
                res = max(res, cur[j]);
            }
            pre = cur;
        }
        return res;
    }
};

File: 718.java
/*
718. Maximum Length of Repeated Subarray
Medium

650

29

Favorite

Share
Given two integer arrays A and B, return the maximum length of an subarray that appears in both arrays.

Example 1:

Input:
A: [1,2,3,2,1]
B: [3,2,1,4,7]
Output: 3
Explanation: 
The repeated subarray with maximum length is [3, 2, 1].
*/

class Solution {
    // dp(i, j) the max subarray end with i and j
    // dp(i, j) = dp(i-1, j-1) + 1 in case of A[i] == B[j] else 0
    public int findLength(int[] A, int[] B) {
        int m = A.length, n = B.length;
        int ans = 0;
        int[] prev = new int[n];
        int[] curr = new int[n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (A[i] != B[j]) curr[j] = 0;
                else if (A[i] == B[j]) curr[j] = j > 0 ? prev[j-1]+1 : 1;
                ans = Math.max(ans, curr[j]);
            }
            int[] t = prev;
            prev = curr;
            curr = t;
        }
        return ans;
    }
}
